### PR TITLE
Support complex number

### PR DESCRIPTION
**Pull request by changhc**
## Change Summary

Support complex numbers.

## Related issue number

Closes https://github.com/pydantic/pydantic/issues/8555

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Also, upgrade to `pydantic-core` v2.22.0, which adds support for complex schemas. We've bundled in some other patches along with this release, such as removing xfails associated with tests that this release fixed!

### CODE REVIEW

File: pydantic/json_schema.py [hunk 1]
```diff
@@ -1847,6 +1847,17 @@ def ser_schema(
...
+    def complex_schema(self, schema: core_schema.ComplexSchema) -> JsonSchemaValue:
+        """Generates a JSON schema that matches a complex number.
+
+        Args:
+            schema: The core schema.
+
+        Returns:
+            The generated JSON schema.
+        """
+        return {'type': 'complex'}
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:1859]: As a reference going forward: https://json-schema.org/understanding-json-schema/reference/numeric

File: tests/test_types.py [hunk 2]
```diff
@@ -6685,3 +6685,49 @@ class Foo(BaseModel):
     # validation error raised bc foo field uses strict mode
     with pytest.raises(ValidationError):
         Foo(foo='1')
+
+
+def test_complex_field():
```
Comment (sydney-runkle) [inline tests/test_types.py:6690]: Tests look good so far, though I think you're right, we'll want to support the string `1+2j` format :)

**Comment by sydney-runkle** [comment]
Hi @changhc,

Great start here - looking good.

Maybe we could add a section to the docs as well with some usage patterns.

**Comment by sydney-runkle** [comment]
I left some notes on next steps in the `pydantic-core` PR :).

**Comment by sydney-runkle** [comment]
@changhc, could you provide an example in `pydantic` where the error type is `complex_str_parsing`?

**Review by sydney-runkle — APPROVED** [review]
Looks good, pending revision of the `complex_str_parsing` error docs in another PR!

**Comment by sydney-runkle** [comment]
Probably could use some more docs here as well. I can handle that and will create an associated issue.

**Comment by changhc** [comment]
Hi @sydney-runkle, I've updated the error doc. The test case that failed with pypy is now skipped. A fix is merged into pypy master branch but not yet released. I'm trying to fix the failed mypy test.

File: tests/mypy/outputs/1.0.1/mypy-plugin_ini/plugin_strict_fields.py [hunk 3]
```diff
@@ -1,56 +1,56 @@
-from pydantic import BaseModel, Field
```
Comment (changhc) [inline tests/mypy/outputs/1.0.1/mypy-plugin_ini/plugin_strict_fields.py:1]: I'm not sure what is actually changed in this file. Everything looks identical. It was "updated" after I ran `make test-mypy-update-all`.

**Comment by sydney-runkle** [comment]
@changhc,

Awesome find! Great work on this feature. We'll be sure to celebrate this in our upcoming release!!

I recall you mentioned you might be interested in working on some other issues. If this is still the case, I can point you towards some fun ones!

**Comment by changhc** [comment]
Thank you for the review! Yes I would love to look into other issues.

By the way, how much does the team care about test coverage? After reading the test coverage report, I'm now adding some tests to cover the missing lines, though I'm not sure if they actually help since the tests to be added are like simply checking if the json schema of a model matches the expected schema.

**Comment by sydney-runkle** [comment]
> By the way, how much does the team care about test coverage? After reading the test coverage report, I'm now adding some tests to cover the missing lines, though I'm not sure if they actually help since the tests to be added are like simply checking if the json schema of a model matches the expected schema.

We care a decent amount about coverage - more tests sound great. Happy to merge after we add those. We generally try to keep coverage above 90%, but closer to 100% would be great.

> Thank you for the review! Yes I would love to look into other issues.

Awesome, I'll ping you with some fun ones soon.

**Comment by sydney-runkle** [comment]
@changhc, as promised, some fun issues:

* https://github.com/pydantic/pydantic/issues/10111
* https://github.com/pydantic/pydantic/issues/9782
* https://github.com/pydantic/pydantic/issues/9677
* https://github.com/pydantic/pydantic/issues/8922
* https://github.com/pydantic/pydantic/issues/8308

**Comment by changhc** [comment]
The missing line in the previous report is now covered.

**Comment by sydney-runkle** [comment]
@changhc, another fun one: https://github.com/pydantic/pydantic/issues/9731

**Pull request merged**